strawberryfields/engine.py

```diff
@@ -531,161 +544,541 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+        """Refreshes the status of the job, along with the job result if the job is
+        newly completed.
+
+        Only an open or queued job can be refreshed; an exception is raised otherwise.
+        """
+        if self.status.is_terminal:
+            raise InvalidJobOperationError(
+                "A {} job cannot be refreshed".format(self.status.value)
+            )
+        self._status = self._connection.get_job_status(self.id)
+        if self._status == JobStatus.COMPLETE:
+            self._result = self._connection.get_job_result(self.id)
+
+    def cancel(self):
+        """Cancels the job.
+
+        Only an open or queued job can be cancelled; an exception is raised otherwise.
+        """
+        if self.status.is_terminal:
+            raise InvalidJobOperationError(
+                "A {} job cannot be cancelled".format(self.status.value)
+            )
+        self._connection.cancel_job(self.id)
+
+    def __repr__(self):
+        return "<{}: id={}, status={}>".format(
+            self.__class__.__name__, self.id, self.status.value
+        )
 
     def __str__(self):
-        return self.__class__.__name__ + "({})".format(self.backend_name)
+        return self.__repr__()
+
+
+class RequestMethod(enum.Enum):
+    """Defines the valid request methods for messages sent to the remote platform."""
+
+    GET = "get"
+    POST = "post"
+    PATCH = "patch"
+
+
+class Connection:
+    """Manages remote connections to the remote job execution platform and exposes
+    various job operations.
+
+    For basic usage, it is not necessary to manually instantiate this object; the user
+    is encouraged to use the higher-level interface provided by :class:`~StarshipEngine`.
+
+    **Example:**
+
+    The following example instantiates a :class:`~Connection` for a given API
+    authentication token, tests the connection, submits a new job, and makes requests
+    for a single or multiple existing jobs.
 
-    def _init_backend(self, *args):
+    .. code-block:: python
+
+        connection = Connection(token="abc")
+
+        # Ping the remote server
+        success = connection.ping()
+        # True if successful, or False if cannot connect or not authenticated
+
+        # Submit a new job
+        job = connection.create_job("chip2", program, shots=123)
+        job  # <Job: id=d177cbf5-1816-4779-802f-cef2c117dc1a, ...>
+
+        # Get all jobs submitted for this token
+        jobs = connection.get_all_jobs()
+        jobs  # [<Job: id=d177cbf5-1816-4779-802f-cef2c117dc1a, ...>, ...]
+
+        # Get a specific job by ID
+        job = connection.get_job("59a1c0b1-c6a7-4f9b-ae37-0ac5eec9c413")
+        job  # <Job: id=59a1c0b1-c6a7-4f9b-ae37-0ac5eec9c413, ...>
+
+    Args:
+        token (str): the API authentication token
+        host (str): the hostname of the remote platform
+        port (int): the port to connect to on the remote host
+        use_ssl (bool): whether to use SSL for the connection
```

Comment:
In the configuration, a `debug` boolean is also specified, I'm missing more details on where it should be incorporated. Maybe it would be intended here as well? :thinking:  @josh146 could you help out with that one?

Reply:
Ah, yes - the former `APIClient` implemented a `debug` mode that simply collected responses and errors in list attributes on the client object. I didn't see a particular use case for that, so left it out for simplicity; but it should not be a very big change to add that in if deemed necessary.

Reply:
I see! Let's see if there would be a use case. Currently, it is included in #298, but that could also be adjusted accordingly. :)

Reply:
Perhaps, rather than debug, a better options would be `verbose=True`. In verbose mode, SF simply prints out details of the job submission to the terminal as it occurs.

@antalszava, might make sense to remove `debug` from #298 for now

Reply:
The `verbose` option has been added!